File: editor/en/[79]Word Search.py
# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def __init__(self):
        self.board = None
        self.rows = None
        self.cols = None

    def exist(self, board: List[List[str]], word: str) -> bool:
        self.board = board
        self.rows = len(board)
        self.cols = len(board[0])
        directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]

        for row in range(self.rows):
            for col in range(self.cols):
                if self.backtrack(row, col, word, 0, directions):
                    return True

        return False

    def backtrack(self, row, col, word, index, directions):
        if index == len(word):
            return True

        if row < 0 or row >= self.rows or col < 0 or col >= self.cols \
                or self.board[row][col] != word[index]:
            return False

        self.board[row][col] = '#'
        for rowOffset, colOffset in directions:
            if self.backtrack(row + rowOffset, col + colOffset, word, index + 1, directions):
                return True

        self.board[row][col] = word[index]
        return False

# leetcode submit region end(Prohibit modification and deletion)

File: editor/leetcode/editor/en/[79]Word Search.java
// leetcode submit region begin(Prohibit modification and deletion)
class Solution {
  private static final int[][] DIRS = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
  private int rows;
  private int cols;

  public boolean exist(char[][] board, String word) {
    rows = board.length;
    cols = board[0].length;

    int[] freqBoard = new int[128];
    for (int r = 0; r < rows; r++) {
      for (int c = 0; c < cols; c++) {
        freqBoard[board[r][c]]++;
      }
    }

    int[] freqWord = new int[128];
    for (char c : word.toCharArray()) {
      freqWord[c]++;
      if (freqWord[c] > freqBoard[c]) {
        return false;
      }
    }

    if (freqBoard[word.charAt(0)] > freqBoard[word.charAt(word.length() - 1)]) {
      word = new StringBuilder(word).reverse().toString();
    }

    for (int r = 0; r < rows; r++) {
      for (int c = 0; c < cols; c++) {
        if (backtrack(board, word, r, c, 0)) {
          return true;
        }
      }
    }

    return false;
  }

  private boolean backtrack(char[][] board, String word, int r, int c, int index) {
    if (index == word.length()) {
      return true;
    }

    if (r < 0 || r >= rows || c < 0 || c >= cols || board[r][c] != word.charAt(index)) {
      return false;
    }

    char temp = board[r][c];
    board[r][c] = '#';

    for (int[] d : DIRS) {
      int nr = r + d[0];
      int nc = c + d[1];

      if (backtrack(board, word, nr, nc, index + 1)) {
        return true;
      }
    }

    board[r][c] = temp;
    return false;
  }
}
// leetcode submit region end(Prohibit modification and deletion)
